packages/engine/src/datastore/table/sync.rs

```diff
@@ -1,15 +1,91 @@
 use std::{fmt, sync::Arc};
 
+use futures::future::join_all;
 use parking_lot::RwLock;
 
 use crate::{
     datastore::{
         prelude::ContextBatch,
         table::pool::{agent::AgentPool, message::MessagePool},
     },
-    worker::runner::comms::inbound::InboundToRunnerMsgPayload,
+    simulation::comms::message::{SyncCompletionReceiver, SyncCompletionSender},
+    worker::{
+        error::{Error as WorkerError, Result as WorkerResult},
+        runner::comms::inbound::InboundToRunnerMsgPayload,
+    },
 };
 
+/// A state sync message with a tokio channel, so that
+/// the sender of the message can wait for the state
+/// sync to complete and find out whether the state
+/// sync succeeded.
+///
+/// Fields:
+/// `completion_sender`: After finishing or failing to sync state,
+///                      used by the receiver/handler of the message
+///                      to notify the sender of the message that the
+///                      state sync completed.
+/// `agent_pool`: Agent batches to load state from
+/// `message_pool`: Message batches to load state from
+#[derive(derive_new::new)]
+pub struct WaitableStateSync {
+    pub completion_sender: SyncCompletionSender,
+    pub agent_pool: AgentPool,
+    pub message_pool: MessagePool,
+}
+
+impl fmt::Debug for WaitableStateSync {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.write_str("WaitableStateSync(...)")
+    }
+}
+
+impl WaitableStateSync {
+    /// Create child messages with the same payload as `self`, which must complete
+    /// for `self` to be complete.
+    pub fn create_children(&self, n_children: usize) -> (Vec<Self>, Vec<SyncCompletionReceiver>) {
+        let mut child_msgs = Vec::new();
+        let mut child_receivers = Vec::new();
+        for _ in 0..n_children {
+            let (sender, receiver) = tokio::sync::oneshot::channel();
+            child_receivers.push(receiver);
+            child_msgs.push(Self {
+                completion_sender: sender,
+                agent_pool: self.agent_pool.clone(),
+                message_pool: self.message_pool.clone(),
+            });
+        }
+        (child_msgs, child_receivers)
+    }
+
+    /// Wait for all child messages to be handled and then send that
+    /// `self` was handled. If any errors occurred while handling a
+    /// child message, send that an error occurred while handling `self`.
+    /// TODO: Return Result (instead of `expect`) and
+    ///       handle where this function is awaited.
+    ///
+    /// Usage:
+    /// let (child_msgs, child_receivers) = self.create_children(2);
+    /// // Send `child_msgs` to appropriate message handlers.
+    /// self.forward_children(child_receivers).await;
```

Comment:
@TimDiekmann should we use a code-block here?

Or would we then have to worry about `self`, and have imports and such

Reply:
I think a code block would be nice, but `self` and such may indeed be a problem. You can flag a code block as `ignore` or `no_run` (e.g. <code>```ignore</code>). See: https://doc.rust-lang.org/rustdoc/documentation-tests.html#attributes. Generally We should try to avoid those attributes, but in the short term it's better than nothing.

Reply:
Yeah makes sense, I've just had a dig through it as well and a bit of a google. Doesn't seem like there are any good approaches really besides introducing a lot of boiler plate and just making it not appear within the docs with `#`.

@litvand feel free to add the codeblock and `ignore` if you like the idea, otherwise I'm fine with it just being plain text I think

Reply:
<code>```text</code> would also be available.